File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        ad = defaultdict(list) #adjacency matrix
        for i in edges:
            ad[i[0]].append(i[1])
            ad[i[1]].append(i[0])
        q = []
        bfs = [0]*(n)
        bfs[source] = 1
        q.append(source)
        while q:
            bfs.append(q[0])
            x = q.pop(0)
            if x==destination:
                return 1
            for i in ad[x]:
                if bfs[i]==0:
                    q.append(i)
                    bfs[i] = 1
        return 0

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {

        List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < n; i++){
            adjList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            adjList.get(edge[0]).add(edge[1]);
            adjList.get(edge[1]).add(edge[0]);
        }
        
        boolean[] visited = new boolean[n];
        return dfs(source, destination, adjList, visited);
    }

    public boolean dfs(int node, int destination, List<List<Integer>> adjList, boolean[] visited) {

        if (node == destination) {
            return true;
        }
        visited[node] = true;

        for (int neighbor : adjList.get(node)) {
            if (!visited[neighbor]){
                if (dfs(neighbor, destination, adjList, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
}